### PR TITLE
Basic transaction implementation

### PR DESCRIPTION
**Pull request by phifty**
A first draft that solves issue #56.

### CODE REVIEW

**Comment by fceller** [comment]
@phifty thanks a lot for your improvements. We are using an Apache 2 CLA for ArangoDB and its companion projects, which can be found here: https://www.arangodb.com/documents/cla.pdf
Just fill in the form, sign and send a scanned copy over to cla[at]arangodb[.]com or via fax +49-221-2722999-88

Note: As we are a company registered in Germany, we are obliged by law to have a signed scanned copy of CLA to be able to accept pull requests to the official ArangoDB repository.

**Comment by phifty** [comment]
OK. @ewoutp said that it's unlikely that this code will be merged. So I would prefer to work on the PR first and if you really want to accept the PR, as the last step, I'm happy to sign the agreement.

**Comment by fceller** [comment]
@phifty yes, I really want to accept the PR. As a better solution will take some time, your solution should be helpful for other people. Still, @ewoutp should give his go on the changes themselves. I'm not a GO programmer.

File: database_impl.go [hunk 1]
```diff
@@ -158,3 +160,41 @@ func (d *database) ValidateQuery(ctx context.Context, query string) error {
...
+		input.Collections.Write = options.WriteCollections
+	}
+	if _, err = req.SetBody(input); err != nil {
+		return nil, WithStack(err)
+	}
+	resp, err := d.conn.Do(ctx, req)
+	if err != nil {
+		return nil, WithStack(err)
+	}
+	if err = resp.CheckStatus(http.StatusOK, http.StatusBadRequest); err != nil {
```
Comment (ewoutp) [inline database_impl.go]: BadRequest should not be in this list.
Note that CheckStatus already knows about the error structure returned by ArangoDB, so no need to parse it yourself later on.

File: database_impl.go [hunk 2]
```diff
@@ -158,3 +160,41 @@ func (d *database) ValidateQuery(ctx context.Context, query string) error {
...
+		return nil, WithStack(err)
+	}
+
+	output := &transactionResponse{}
+	if err = resp.ParseBody("", output); err != nil {
+		return nil, WithStack(err)
+	}
+
+	if output.Error {
```
Comment (ewoutp) [inline database_impl.go]: Please remove this. Is part of CheckStatus.

File: example_create_document_test.go [hunk 3]
```diff
@@ -54,6 +54,7 @@ func Example_createDocument() {
 	if err != nil {
 		log.Fatalf("Failed to create database: %v", err)
 	}
+	defer db.Remove(nil)
```
Comment (ewoutp) [inline example_create_document_test.go]: Please remove this change

File: environment_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package driver_test
```
Comment (ewoutp) [inline environment_test.go]: Why is this file needed

File: transaction.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+// TransactionOptions contains options that customize the transaction.
+type TransactionOptions struct {
+	// Transaction size limit in bytes. Honored by the RocksDB storage engine only.
+	MaxTransactionSize int
+
+	// An optional numeric value that can be used to set a timeout for waiting on collection
+	// locks. If not specified, a default value will be used.
+	// Setting lockTimeout to 0 will make ArangoDB not time out waiting for a lock.
+	LockTimeout interface{}
```
Comment (ewoutp) [inline transaction.go]: Why `interface{}`? 
Probably you want to use and `*int`

File: transaction.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	// An optional boolean flag that, if set, will force the transaction to write
+	// all data to disk before returning.
+	WaitForSync bool
+
+	// Maximum number of operations after which an intermediate commit is performed
+	// automatically. Honored by the RocksDB storage engine only.
+	IntermediateCommitCount interface{}
```
Comment (ewoutp) [inline transaction.go]: `*int` ??

File: transaction.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+	// Optional arguments passed to action.
+	Params []interface{}
+
+	// Maximum total size of operations after which an intermediate commit is
+	// performed automatically. Honored by the RocksDB storage engine only.
+	IntermediateCommitSize interface{}
```
Comment (ewoutp) [inline transaction.go]: `*int`?

File: transaction.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+type transactionCollectionsRequest struct {
+	Read  []string `json:"read,omitempty"`
+	Write []string `json:"write,omitempty"`
+}
+
+type transactionResponse struct {
+	Error        bool        `json:"error"`
+	Code         int         `json:"code"`
+	Result       interface{} `json:"result"`
```
Comment (ewoutp) [inline transaction.go]: This is the only field needed in here. Rest is handled in CheckStatus

File: transaction.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	IntermediateCommitSize interface{}
+
+	// Collections that the transaction reads from.
+	ReadCollections []string
+
+	// Collections that the transaction writes to.
+	WriteCollections []string
+}
+
+type transactionRequest struct {
```
Comment (ewoutp) [inline transaction.go]: Please use appropriate types: `*int` probably

File: database_test.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package driver_test
```
Comment (ewoutp) [inline database_test.go]: Suggest to name this `database_transaction_test.go`

File: database_test.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	driver "github.com/arangodb/go-driver"
+	"github.com/stretchr/testify/assert"
```
Comment (ewoutp) [inline database_test.go]: Please remove the `github.com/stretchr/testify` dependency.
We want this driver to be as self-containing as possible

File: database_test.go (hunk 10 above)
Comment (ewoutp) [reply database_test.go]: and move it to test folder.

File: example_create_document_test.go (hunk 3 above)
Comment (phifty) [reply example_create_document_test.go]: Ok, why is this test existing? It does not run more than one time.

File: transaction.go (hunk 5 above)
Comment (phifty) [reply transaction.go]: I wanted to be able to write `TransactionOptions{LockTimeout: 5}` instead of

    lockTimeout := 5
    TransactionOptions{LockTimeout: &lockTimeout}

but if type-safety is not a valid trade-off at this point, I can change that.

File: environment_test.go (hunk 4 above)
Comment (phifty) [reply environment_test.go]: I want the test to be focused on the thing it tests. So I moved the setup/teardown code to this file.

File: example_create_document_test.go (hunk 3 above)
Comment (ewoutp) [reply example_create_document_test.go]: This is not really a test. It is example code which needs to have a name such as this to avoid inclusion in the normal code.

File: environment_test.go (hunk 4 above)
Comment (ewoutp) [reply environment_test.go]: Since all test code is in the test folder and setup is also in there, I don't want to pollute the root folder with separate test code.

File: transaction.go (hunk 5 above)
Comment (ewoutp) [reply transaction.go]: I understand the argument, but I prefer type safety over ease of use in this case.

File: example_create_document_test.go (hunk 3 above)
Comment (phifty) [reply example_create_document_test.go]: That seems a bit odd to me. The tests - that I usually see in the `*_test` package - are placed in the `test` folder. And the examples - that I usually see in the `example` folder - are in the `*_test` package :-)

Anyway - that exceeds the scope of this PR and should not concern me.

**Comment by phifty** [comment]
I'd be done with my changes so far.

**Comment by fceller** [comment]
@phifty thanks a lot for your work.

@ewoutp please let us know if you are fine with changes. We need a CLA before we can merge.

File: database_impl.go [hunk 12]
```diff
@@ -158,3 +159,38 @@ func (d *database) ValidateQuery(ctx context.Context, query string) error {
...
+		input.MaxTransactionSize = options.MaxTransactionSize
+		input.LockTimeout = options.LockTimeout
+		input.WaitForSync = options.WaitForSync
+		input.IntermediateCommitCount = options.IntermediateCommitCount
+		input.Params = options.Params
+		input.IntermediateCommitSize = options.IntermediateCommitSize
+		input.Collections.Read = options.ReadCollections
+		input.Collections.Write = options.WriteCollections
+	}
+	if _, err = req.SetBody(input); err != nil {
```
Comment (ewoutp) [inline database_impl.go]: When using an `if` with a nested variable lik this, I prefer to have the nested variable defined explicitly.
Like:

`if _, err := req.SetBody(input); err != nil {...`

Feel free to change. This is not a show stopper.

**Comment by ewoutp** [comment]
@phifty Thanks for your changes. 
LGTM.

@fceller Once I get the confirmation that the CLA is done, I'll merge it.

**Comment by fceller** [comment]
@ewoutp CLA is available, merge at will
